src/NITTA/Model/Networks/Bus.hs

```diff
@@ -733,3 +777,77 @@ instance (UnitTag tag, VarValTime v x t) => Testable (BusNetwork tag v x t) v x
+
+data BuilderSt tag v x t = BuilderSt
+    { signalBusWidth :: Int
+    , availSignals :: [SignalTag]
+    , pus :: M.Map tag (PU v x t)
+    , reserve :: M.Map tag (PUTemplate v x t)
+    }
+
+modifyNetwork net@BusNetwork{bnPus, bnPUReserve, bnSignalBusWidth, bnEnv} builder =
+    let st0 =
+            BuilderSt
+                { signalBusWidth = bnSignalBusWidth
+                , availSignals = map (SignalTag . controlSignalLiteral) [bnSignalBusWidth :: Int ..]
+                , pus = bnPus
+                , reserve = bnPUReserve
+                }
+        BuilderSt{signalBusWidth, pus, reserve} = execState builder st0
+        netIOPorts ps =
+            BusNetworkIO
+                { extInputs = unionsMap puInputPorts ps
+                , extOutputs = unionsMap puOutputPorts ps
+                , extInOuts = unionsMap puInOutPorts ps
+                }
+     in net
+            { bnPus = pus
+            , bnSignalBusWidth = signalBusWidth
+            , bnEnv = bnEnv{ioPorts = Just $ netIOPorts $ M.elems pus}
+            , bnPUReserve = reserve
+            }
+
+defineNetwork bnName ioSync builder = modifyNetwork (busNetwork bnName ioSync) builder
+
+addCustom tag pu ioPorts = do
+    st@BuilderSt{signalBusWidth, availSignals, pus} <- get
+    let ctrlPorts = takePortTags availSignals pu
+        puEnv =
+            def
+                { ctrlPorts = Just ctrlPorts
+                , ioPorts = Just ioPorts
+                , valueIn = Just ("data_bus", "attr_bus")
+                , valueOut = Just (toText tag <> "_data_out", toText tag <> "_attr_out")
+                }
+        pu' = PU pu def puEnv
+        usedPortsLen = length $ usedPortTags ctrlPorts
+    put
+        st
+            { signalBusWidth = signalBusWidth + usedPortsLen
+            , availSignals = drop usedPortsLen availSignals
+            , pus =
+                if M.member tag pus
+                    then error "every PU must has uniq tag"
+                    else M.insert tag pu' pus
+            }
+
+-- |Add PU with the default initial state. Type specify by IOPorts.
+add tag ioport = addCustom tag def ioport
+
+addCustomToReserve :: forall tag v x t m pu. (MonadState (BuilderSt tag v x t) m, PUClasses pu v x t, UnitTag tag) => tag -> pu -> IOPorts pu -> m ()
+addCustomToReserve tag pu ioports
+    | typeOf pu == typeRep (Proxy :: Proxy (SPI v x t)) = error "Adding SPI to reserve are not supported due to https://github.com/ryukzak/nitta/issues/194"
+    | otherwise = do
+        st@BuilderSt{reserve} <- get
+        put
+            st
+                { reserve =
+                    if M.member tag reserve
+                        then error "every PU in reserve must has uniq tag"
+                        else M.insert tag (PUTemplate pu ioports) reserve
+                }
+
+-- |Add PU to reserve with the default initial state. Type specify by IOPorts.
+addToReserve tag ioports = addCustomToReserve tag def ioports
```

Comment:
A very good question is how it should be done…. Only one idea of what I can provide right now — we need one place with the default value for reserved list.

Also — reserve not a good term. I think the prototype will be better.